From b7a0b3f719a75ec60b74450943dd8ef36b541bbf Mon Sep 17 00:00:00 2001
From: Wei Qin <qinwei61@huawei.com>
Date: Fri, 20 Mar 2026 09:17:49 +0800
Subject: [PATCH 1/1] ub: udma: bugfix related to clean jfc.

driver inclusion
category: bugfix
bugzilla: https://atomgit.com/openeuler/umdk/issues/196
CVE: NA

-----------------------------------------------------------

This patch fix a bug related to clean jfc.

Fixes: a7ea7e683ec6 ("udma: support alloc and free tid")

Signed-off-by: Wei Qin <qinwei61@huawei.com>
---
 src/urma/hw/udma/udma_u_jfc.c | 12 ++++++++++--
 src/urma/hw/udma/udma_u_jfr.c |  2 ++
 src/urma/hw/udma/udma_u_jfs.c |  1 -
 3 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/src/urma/hw/udma/udma_u_jfc.c b/src/urma/hw/udma/udma_u_jfc.c
index 166ff43..86552cf 100644
--- a/src/urma/hw/udma/udma_u_jfc.c
+++ b/src/urma/hw/udma/udma_u_jfc.c
@@ -72,6 +72,7 @@ static int udma_u_create_cq(struct udma_u_jetty_queue *cq, urma_jfc_cfg_t *cfg)
 {
 	uint32_t depth;
 
+	cq->ci = 0;
 	cq->lock_free = cfg->flag.bs.lock_free;
 	if (!cq->lock_free &&
 	    pthread_spin_init(&cq->lock, PTHREAD_PROCESS_PRIVATE)) {
@@ -220,10 +221,16 @@ err_alloc_sw_db:
 static int udma_u_set_jfc_depth(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
 {
 	urma_cmd_udrv_priv_t udata = {};
+	int ret;
 
+	ret = urma_cmd_set_jfc_opt(jfc, opt, buf, len, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("fail to set jfc depth, depth = %u.\n", *(uint32_t *)buf);
+		return ret;
+	}
 	jfc->jfc_cfg.depth = *(uint32_t *)buf;
 
-	return urma_cmd_set_jfc_opt(jfc, opt, buf, len, &udata);
+	return ret;
 }
 
 static int udma_u_set_jfc_ceqn(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
@@ -922,7 +929,8 @@ void udma_u_clean_jfc(struct urma_jfc *jfc, uint32_t jetty_id)
 	urma_cr_t cr;
 	uint32_t pi;
 
-	if (udma_u_jfc->mode != (uint32_t)UDMA_U_NORMAL_JFC_TYPE)
+	if ((udma_u_jfc->mode != (uint32_t)UDMA_U_NORMAL_JFC_TYPE) ||
+	    (udma_u_jfc->cq.cstm))
 		return;
 
 	cq = &udma_u_jfc->cq;
diff --git a/src/urma/hw/udma/udma_u_jfr.c b/src/urma/hw/udma/udma_u_jfr.c
index 4f99e71..b96a770 100644
--- a/src/urma/hw/udma/udma_u_jfr.c
+++ b/src/urma/hw/udma/udma_u_jfr.c
@@ -47,6 +47,8 @@ void udma_u_init_jfr_param(struct udma_u_jfr *jfr, urma_jfr_cfg_t *cfg)
 	jfr->wqe_shift = UDMA_U_ILOG32(roundup_pow_of_two(UDMA_SGE_SIZE *
 							  jfr->max_sge));
 	jfr->lock_free = cfg->flag.bs.lock_free;
+	jfr->rq.pi = 0;
+	jfr->rq.ci = 0;
 }
 
 static int udma_u_create_rq(struct udma_u_context *udma_ctx,
diff --git a/src/urma/hw/udma/udma_u_jfs.c b/src/urma/hw/udma/udma_u_jfs.c
index d6e3a76..19bca9e 100644
--- a/src/urma/hw/udma/udma_u_jfs.c
+++ b/src/urma/hw/udma/udma_u_jfs.c
@@ -297,7 +297,6 @@ err_create_sq:
 
 static void udma_u_free_jfs_prepare(urma_jfs_t *jfs)
 {
-	struct udma_u_context *udma_ctx = to_udma_u_ctx(jfs->urma_ctx);
 	struct udma_u_jfs *udma_jfs = to_udma_u_jfs(jfs);
 
 	if (!!jfs->jfs_cfg.jfc)
-- 
2.28.0.windows.1

